src/plugins/spec/spec-new.c

```diff
@@ -0,0 +1,762 @@
+	strcpy (copiedKeyName, (char *) specKeyName);
+
+	char * rest = NULL;
+	char * arrayParent = strtok_r (copiedKeyName, "#", &rest);
+	arrayParent[strlen (arrayParent) - 1] = '\0';
+
+	Key * key = getMatchingKeyFromKeySet (specKeys, strchr (arrayParent, '/'));
+
+	elektraFree (copiedKeyName);
+
+	return key;
+}
+
+/**
+ * Get the number of array characters in an array specification key name.
+ *
+ * @param specKey the specification key to use
+ * @return number of array characters in the array specification key name
+ */
+static int getNumberOfArrayCharactersInSpecName (Key * specKey)
+{
+	char * withoutNamespace = strchr (keyName (specKey), '/');
+
+	int count = 0;
+
+	for (size_t i = 0; i < elektraStrLen (withoutNamespace); i++)
+	{
+		if (withoutNamespace[i] == '#')
+		{
+			count++;
+		}
+	}
+
+	return count;
+}
+
+/**
+ * The positions of where the `#` occurs in the specification key.
+ *
+ * @param keyNameWithoutNamespace the key name without the namespace
+ * @param arrayPositions the array to hold all the positions of the `#` in the specification key
+ * @param arraySize number of `#` in the whole specification key
+ */
+static void setArrayPositions (const char * keyNameWithoutNamespace, int * arrayPositions)
+{
+	int arrPos = 0;
+	for (int i = 0; i < (int) elektraStrLen (keyNameWithoutNamespace); i++)
+	{
+		if (keyNameWithoutNamespace[i] == '#')
+		{
+			arrayPositions[arrPos++] = i;
+		}
+	}
+}
+
+/**
+ * Validate the array size. If it does not find any array size a warning is appended to the parent key.
+ *
+ * @param ks the key set to look for the key
+ * @param specKeys the specification keys
+ * @param specKey the array specification key
+ * @retval true - if the array size is valid
+ * @retval false - if the array size was not found
+ * @retval false - if the array size is not valid
+ */
+static bool isValidArraySize (KeySet * ks, KeySet * specKeys, Key * specKey)
+{
+	Key * key = ksLookupByName (ks, strchr (keyName (specKey), '/'), 0);
+	Key * keyToFetchArraySizeFrom = key == NULL ? getArraySizeOfArrayParent (specKeys, specKey) : key;
+	const Key * arrayMetaKey = key == NULL ? keyGetMeta (keyToFetchArraySizeFrom, "array") : keyGetMeta (key, "array");
+
+	if (arrayMetaKey == 0)
+	{
+		// no array size found, skip
+		return true;
+	}
+
+	if (!validateArraySize (keyToFetchArraySizeFrom, specKey))
+	{
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * Check if the array is empty.
+ *
+ * @param ks the KeySet with all keys
+ * @param arrayPosition the position of the array character
+ * @retval true - if the array is empty
+ * @retval false - if the array is not empty
+ */
+static bool isArrayEmpty (KeySet * ks, int arrayPosition)
+{
+	for (elektraCursor it = 0; it < ksGetSize (ks); it++)
+	{
+		Key * current = ksAtCursor (ks, it);
+		char * withoutNamespace = strchr (keyName (current), '/');
+
+		int len = elektraStrLen (withoutNamespace);
+		if (arrayPosition > len)
```

Comment:
```suggestion
		if (arrayPosition >= (len-1))
```
`len` are the numbers of characters including `\0`, so the last valid index is `len-1` where the `\0` has to be.
The last possible position for the array character is therefore at `len-2`.

Reply:
Isn't `> len` equal to `>= len-1`?

Reply:
@atmaxinger The other way around. If `>= len+1`, just `> len` would be fine.